### ISSUE TITLE
Mismatch between assertion string and implementation

### ISSUE THREAD

**Issue by Erik-Lundell** [pytorch/executorch#7782]
Hi! In `exir.program._program._get_updated_graph_signature` that updates an ExportGraphSignature with a new GraphModule there is a an assert that checks that the number of placeholders in the new GraphModule is less than the number of input compile specs in the ExportGraphSignature:

https://github.com/pytorch/executorch/blob/8611aaa064d1bff31a0f13850c6c0f66608895a4/exir/program/_program.py#L156C9-L158C28 

However, the assert string says that the number of inputs has *changed*, not increased. The implementation also seems to assume that the inputs are in the same order (i.e. no removed placeholders in the middle of the list).

Does this mean actually mean that a transform of a  GraphModule should not change the inputs (and outputs) or is the implementation flimsy? Is there a reason for not changing the lifted inputs when transforming a graph_module? 

To provide broader context, I am implementing a fuse pass (conv + batchnorm) that removes some placeholder nodes, similar to https://github.com/pytorch/executorch/blob/main/backends/transforms/fuse_batch_norm_with_conv.py except that I update the conv weights and bias instead of adding get_attr nodes. `_get_updated_graph_signature` is then called via `exir.program._transform`.

Since I remove placeholder nodes corresponding to elements in the middle of the InputSpecs list, the `_get_update_graph_signature` function mismatches the old InputSpec list with the new placeholder nodes. How can I avoid that? The other pass I mentioned seems to just skip removing the unused nodes but that also seems awkward?

cc @digantdesai @freddan80 @per @zingo @oscarandersson8218

**Comment by digantdesai** [comment]
Thanks @Erik-Lundell  for the detailed description with useful pointers. And apologies for the delay.

> Does this mean actually mean that a transform of a GraphModule should not change the inputs (and outputs) 

I am not sure. But that doesn't seem like very useful/meaningful constraint. If I were to guess the `_transform()` helper function seems like written with this assumption. However, you don't have to use that, if you look at `lift_constant_tensor_pass` which doesn't use that, in the same file, and perhaps use that as an example to update the signature as you see fit.

For the `fuse_batch_norm_with_conv` pass - it does add attributes for fused weights/biases, which should be OK inside the delegate given we won't run edge dialect verifier on it. XNNPACK does that today, in addition to other violations. The rationale is that graph isn't going to be further lowered with edge IR constraints, and going to be consumed by the delegate. ET might looks at it after to_backend, and if you go too far something might break but for small rule-breaks it should be acceptable.

**Comment by Erik-Lundell** [comment]
Interesting, thanks for your comments. I decided to go with using the existing placeholders in the end but it's useful to know that it should be possible to update the graph_signature if needed. Thanks!
